# Homework 5

## Tyler Valdez (UTEID: trv359)

## Logistic Regression

In [1]:
# Standard Headers
# You are welcome to add additional headers here if you wish
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.datasets import make_blobs
from sklearn.metrics import classification_report

from matplotlib import pyplot
import matplotlib.cm as cm
from random import sample

# Enable inline mode for matplotlib so that Jupyter displays graphs
%matplotlib inline

## Heart Dataset 

In this Assignment we will work with some patients dataset. 

We have access to 303 patients data. The features are listed below. 

In [2]:
# Your code here
heart_df = pd.read_csv("Heart.csv")
heart_df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal Target  
0        2.3      3  0.0       fixed     No  
1        1.5      2  3.0      normal    Yes  
2        2.6      2  2.0  reversable    Yes  
3        3.5      3  0.0      normal     No  
4        1.4      1  0.0      normal     No  
..       ...    ...  ...         ...    ...  
298      1.2      2  0.0  reversable    Yes  
299      3.4      2  2.0  reversable    Yes  
300      1.2      2  1.0  reversable    Yes  
301      0.0      2  1.0      normal    Yes  
302      0.0      1  NaN      normal     No  

[303 rows x 14 columns]

**Age:** The person’s age in years

**Sex:** The person’s sex (1 = male, 0 = female)

**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

**RestBP:** The person’s resting blood pressure (mm Hg on admission to the hospital)

**Chol:** The person’s cholesterol measurement in mg/dl

**Fbs:** The person’s fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)
restecg: resting electrocardiographic results

* Value 0: showing probable or definite left ventricular hypertrophy by Estes’ criteria
* Value 1: normal
* Value 2: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)

**RestECG:** The person’s maximum heart rate achieved

**MaxHR:** Exercise induced angina (1 = yes; 0 = no)

**Oldpeak:** ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot. See more here)

**Slope:** the slope of the peak exercise ST segment — 0: downsloping; 1: flat; 2: upsloping

* 0: downsloping; 
* 1: flat; 
* 2: upsloping

**Ca:** The number of major vessels (0–3)

**Thal:** A blood disorder called thalassemia Value 0: NULL (dropped from the dataset previously

* Value 1: fixed defect (no blood flow in some part of the heart)
* Value 2: normal blood flow
* Value 3: reversible defect (a blood flow is observed but it is not normal)

**Target:** Heart disease (1 = no, 0= yes)

# Task - 1 (5 points)
We want to use logistic regerssion to perdict if the patients will have heart problems or not. The column "Target" in our datasets includes data about heart diseases. If the patient had heart disease we have a 1 and if not a zero. 

Prepare your data set for predicting heart disease ("Target" column) out of 3 features:

* Age of the patient (Column **"Age"**)
* Gender of the patient (male or female - Column **"Sex"**)
* Cholestrol level of the patient (Column **"Chol"**) 


Split your data into 80% traning data and 20% test data. 

In [24]:
# Split dataset in features and target variable
feature_cols = ['Age', 'Sex', 'Chol']
X = heart_df[feature_cols] # Features
y = heart_df['Target'] # Target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state=0)

print(X_train, X_test)
print(y_train, y_test)

     Age  Sex  Chol
74    44    1   197
153   55    1   289
64    54    1   188
296   59    1   176
287   58    1   220
..   ...  ...   ...
251   58    1   218
192   43    1   247
117   35    0   183
47    50    1   243
172   59    0   249

[242 rows x 3 columns]      Age  Sex  Chol
225   34    0   210
152   67    0   564
228   54    1   206
201   64    0   325
52    44    1   290
..   ...  ...   ...
146   57    1   289
302   38    1   175
26    58    0   340
108   61    1   260
89    51    0   256

[61 rows x 3 columns]
74     Yes
153    Yes
64     Yes
296    Yes
287     No
      ... 
251    Yes
192    Yes
117     No
47     Yes
172    Yes
Name: Target, Length: 242, dtype: object 225     No
152     No
228    Yes
201     No
52     Yes
      ... 
146    Yes
302     No
26      No
108    Yes
89      No
Name: Target, Length: 61, dtype: object


# Task 2 - (10 points)

Generate a logistic regression model using your training data. 

Print out the accuracy if your model. 

In [27]:
# Building and Training a Logistic Regression Model 
model = LogisticRegression(penalty='none', fit_intercept=False)
model.fit(X_train,y_train)
y_pred = model.predict(X_test)
# Prints out accuracy of Logistic Regression Model
print("The accuracy score of my Logistic Regression Model is",accuracy_score(y_test,y_pred))


The accuracy score of my Logistic Regression Model is 0.7049180327868853


# Task 3 - (5 points)


Generate the classification report for your Logistic regresion model and interpret your results regarding precision, recall and f1-score. 


**Precision - Out of all the patients that the model predicted would have heart disease only, 61% actually did.**

**Recall - Out of all the patients that did have heart disease, the model only predicted this outcome correctly for 85% of the patients.**

**F1-Score - Since the F1 value is close to 1, it tells us this model did a good job of predicting whether or not patients will have heart disease or not.**

In [28]:
# Using the result of my accuracy score 
# Print out a classification report of my model
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

          No       0.84      0.60      0.70        35
         Yes       0.61      0.85      0.71        26

    accuracy                           0.70        61
   macro avg       0.73      0.72      0.70        61
weighted avg       0.74      0.70      0.70        61



# Task 4 - Optional Task (2 extra points)


Which other feature of your data can you use to improve your prediction? Build other models and improve your prediction. 
